stan/math/fwd/fun/multiply.hpp

```diff
@@ -10,120 +11,50 @@
-    const Eigen::Matrix<fvar<T>, R1, C1>& m1,
-    const Eigen::Matrix<fvar<T>, R2, C2>& m2) {
+template <typename T1, typename T2,
+          typename = require_all_eigen_vt<is_fvar, T1, T2>,
+          typename = require_same_vt<T1, T2>,
+          typename = require_not_eigen_row_and_col_t<T1, T2>, unsigned int = 0>
+inline auto multiply(const T1& m1, const T2& m2) {
   check_multiplicable("multiply", "m1", m1, "m2", m2);
-  Eigen::Matrix<fvar<T>, R1, C2> result(m1.rows(), m2.cols());
-  for (size_type i = 0; i < m1.rows(); i++) {
-    Eigen::Matrix<fvar<T>, 1, C1> crow = m1.row(i);
-    for (size_type j = 0; j < m2.cols(); j++) {
-      Eigen::Matrix<fvar<T>, R2, 1> ccol = m2.col(j);
-      result(i, j) = dot_product(crow, ccol);
-    }
-  }
-  return result;
+  return m1 * m2;
 }
 
-template <typename T, int R1, int C1, int R2, int C2>
-inline Eigen::Matrix<fvar<T>, R1, C2> multiply(
-    const Eigen::Matrix<fvar<T>, R1, C1>& m1,
-    const Eigen::Matrix<double, R2, C2>& m2) {
+template <typename T1, typename T2, typename = require_eigen_vt<is_fvar, T1>,
+          typename = require_eigen_vt<std::is_floating_point, T2>,
+          typename = require_not_eigen_row_and_col_t<T1, T2>, int = 0>
+inline auto multiply(const T1& m1, const T2& m2) {
   check_multiplicable("multiply", "m1", m1, "m2", m2);
-  Eigen::Matrix<fvar<T>, R1, C2> result(m1.rows(), m2.cols());
+  Eigen::Matrix<value_type_t<T1>, T1::RowsAtCompileTime, T2::ColsAtCompileTime>
+      result(m1.rows(), m2.cols());
   for (size_type i = 0; i < m1.rows(); i++) {
-    Eigen::Matrix<fvar<T>, 1, C1> crow = m1.row(i);
+    Eigen::Matrix<value_type_t<T1>, 1, T1::ColsAtCompileTime> crow = m1.row(i);
     for (size_type j = 0; j < m2.cols(); j++) {
-      Eigen::Matrix<double, R2, 1> ccol = m2.col(j);
+      auto ccol = m2.col(j);
       result(i, j) = dot_product(crow, ccol);
     }
   }
   return result;
 }
 
-template <typename T, int R1, int C1, int R2, int C2>
-inline Eigen::Matrix<fvar<T>, R1, C2> multiply(
-    const Eigen::Matrix<double, R1, C1>& m1,
-    const Eigen::Matrix<fvar<T>, R2, C2>& m2) {
+template <typename T1, typename T2,
+          typename = require_eigen_vt<std::is_floating_point, T1>,
+          typename = require_eigen_vt<is_fvar, T2>,
+          typename = require_not_eigen_row_and_col_t<T1, T2>, char = 0>
+inline auto multiply(const T1& m1, const T2& m2) {
   check_multiplicable("multiply", "m1", m1, "m2", m2);
-  Eigen::Matrix<fvar<T>, R1, C2> result(m1.rows(), m2.cols());
+  Eigen::Matrix<value_type_t<T2>, T1::RowsAtCompileTime, T2::ColsAtCompileTime>
+      result(m1.rows(), m2.cols());
   for (size_type i = 0; i < m1.rows(); i++) {
-    Eigen::Matrix<double, 1, C1> crow = m1.row(i);
+    Eigen::Matrix<double, 1, T1::ColsAtCompileTime> crow = m1.row(i);
     for (size_type j = 0; j < m2.cols(); j++) {
-      Eigen::Matrix<fvar<T>, R2, 1> ccol = m2.col(j);
+      auto ccol = m2.col(j);
       result(i, j) = dot_product(crow, ccol);
     }
   }
```

Comment:
This is a case where less is not better.

Doing `Eigen::Matrix<double, 1, T1::ColsAtCompileTime> crow = m1.row(i);` in outer loop benefits performance for large matrices. Consecutive elements in same row of the matrix are not in consecutive memory locations, so accessing them in order uses cache poorly. Copying them to a vector makes them consecutive in memory and accessing them faster. 

Since we go over them in every iteration of inner loop it is better to inefficiently access them just once and make remaining accesses efficient.

If you want I can make line 51 na 52 into a single line, but it won't affect performance in any way.

Reply:
> Doing Eigen::Matrix<double, 1, T1::ColsAtCompileTime> crow = m1.row(i); in outer loop benefits performance for large matrices. Consecutive elements in same row of the matrix are not in consecutive memory locations, so accessing them in order uses cache poorly. Copying them to a vector makes them consecutive in memory and accessing them faster.

> Since we go over them in every iteration of inner loop it is better to inefficiently access them just once and make remaining accesses efficient.

Word that's sensible to me.

> If you want I can make line 51 na 52 into a single line, but it won't affect performance in any way.

idt it will effect performance though I think zipping 51 into 52 would be nice

Reply:
done